src/sdc11073/roles/waveformprovider/waveforms.py

```diff
@@ -0,0 +1,70 @@
+import itertools
+import math
+from typing import Callable
+
+curve_generator = Callable[[float, float, int], list[float]]
+
+
+def sinus(min_value: float, max_value: float, samples: int) -> list[float]:
+    """Return a list of values for one sinus curve period."""
+    delta = 2 * math.pi / samples
+    _values = [math.sin(i * delta) for i in range(samples)]  # -1 ... +1
+    return [(n + 1) / 2.0 * (max_value - min_value) + min_value for n in _values]  # min ... max
+
+
+def sawtooth(min_value: float, max_value: float, samples: int) -> list[float]:
+    """Return a list of values for one sawtooth curve period."""
+    delta = (max_value - min_value) / float(samples)
+    return [min_value + i * delta for i in range(samples)]
+
+
+def triangle(min_value: float, max_value: float, samples: int) -> list[float]:
+    """Return a list of values for one triangle curve period."""
+    min_value = float(min_value)
+    max_value = float(max_value)
+    delta = (max_value - min_value) / float(samples) * 2
+    samples_cnt = int(samples / 2)
+    return [min_value + i * delta for i in range(samples_cnt)] + [max_value - i * delta for i in range(samples_cnt)]
+
+
+class WaveformGeneratorBase:
+    """Generator of infinite curve, data is provided by a curve generator."""
+
+    def __init__(self,
+                 values_generator: curve_generator,
+                 min_value: float,
+                 max_value: float,
+                 waveform_period: float,
+                 sample_period: float):
+        if sample_period >= waveform_period:
+            raise ValueError(
+                f'please choose a waveformperiod >> sampleperiod. currently use have wp={waveform_period}, sp={sample_period}')
+        self.sample_period = sample_period
+        samples = int(waveform_period / sample_period)
```

Comment:
this fails if `sample_period` is `0`. maybe it is better to add a check with a more explanatory message than `ZeroDivisionError: division by zero`

Reply:
added a check for <= 0